GPS.device.nut

```diff
@@ -0,0 +1,258 @@
+                    local lat = parsedFields[1];
+                    if (lat.len() <= 1) break;
+                    
+                    _extractLat(lat, retTable, parsedFields[2]);
+                    _extractLong(parsedFields[3], retTable, parsedFields[4]);
+                    _extractTime(parsedFields[5], retTable);
+                    
+                    retTable.status <- (parsedFields[6] == "A" ? "Active" : "Void");
+                    break;
+                // Essential Fix data
+                case GPS_GGA:
+                    retTable.type <- GPS_GGA;
+                    local lat = parsedFields[2];
+                    if (lat.len() <= 1) break; // no data
+                    
+                    _extractTime(parsedFields[1], retTable);
+                    _extractLat(lat, retTable, parsedFields[3]);
+                    _extractLong(parsedFields[4], retTable, parsedFields[5]);
+                    
+                    retTable.status <- "Active";
+                    retTable.fixQuality <- parsedFields[6];
+                    retTable.numSatellites <- parsedFields[7].tointeger();
+                    retTable.altitude <- parsedFields[9].tofloat();
+                    break;
+                // Satellites in view
+                case GPS_GSV:
+                    retTable.type <- GPS_GSV;
+                    retTable.numSatellites <- parsedFields[3].tointeger();
+                    break;
+                // GPS DOP and active satellites
+                case GPS_GSA:
+                    retTable.type <- GPS_GSA;
+                    retTable.threeDFix <- parsedFields[2];
+                    local len = parsedFields.len();
+                    // need to use the len b/c we don't know the number of satellite PRNs
+                    retTable.PDOP <- parsedFields[len-4];
+                    retTable.HDOP <- parsedFields[len-3];
+                    retTable.VDOP <- parsedFields[len-2];
+                    break;
+                // return empty table if not a supported data type
+                default:
+                    return {};
+            }
+        }
+        local checkLen = parsedFields[parsedFields.len() - 1].len();
+        retTable.checkSum <- parsedFields[parsedFields.len()-1].slice(checkLen-4, checkLen-2);
+        return retTable;
+    }
+
+    // Get the latitude from strings. Store it in tb
+    function _extractLat(str, tb, direction) {
+        local lat = str.slice(0, 2).tofloat() + (str.slice(2).tofloat()/60);
+        if (direction == "S") lat = -lat;
+        tb.latitude <- lat;
+    }
+
+    // Get the longitude from strings. Store it in tb
+    function _extractLong(str, tb, direction) {
+        local long = str.slice(0, 3).tofloat() + (str.slice(3).tofloat()/60);
+        if (direction == "W") long = -long;
+        tb.longitude <- long;
+    }
+
+    // Get the time from a string. Store it in tb
+    function _extractTime(str, tb) {
+        local time = str.tointeger();
```

Comment:
It would be better to return the time (either as a table or just the epoch time stamp?) rather than writing time keys to the table that is passed in.